### PR TITLE
Feature bsc

### PR DESCRIPTION
**Pull request by idirall22**
# New BSC Feature

The new BSC feature allows Rainbow bridge to run on Binance Smart Chain.

# Technical

When the Bridge is configured to use **bsc** mode, the first header has to be an epoch header (block_number%200==0). That is because the epoch header contains the validator set inside the `extra_data` field.  After that, bridge starts to receive and validate the other blocks. If current block at the time of starting the bridge isn't the epoch block, bridge gets initialised at previous epoch block.
Every time Near client gets a new epoch header it saves it and uses the new validator set from it.

# New Configurations:

The Bridge handles POW(ethash), POSA(bsc), or POA, by default the POW is used.

To use `bsc`:
```json
    nearClientValidateHeader: 'true',
    nearClientValidateHeaderMode: 'bsc'
```

and to use `ethash`:
```json
    nearClientValidateHeader: 'true',
    nearClientValidateHeaderMode: 'ethash'
```

here is an example of the new config file `~/.rainbow/config.json`:
```json
    {
        "nearNodeUrl": "http://localhost:3030",
        "nearNetworkId": "local",
        "nearMasterAccount": "node0",
        "nearMasterSk": "ed25519:3D4YudUQRE39Lc4JHghuB5WM8kbgDDa34mnrEP5DdTApVH81af7e2dWgNPEaiQfdJnZq1CNPp5im4Rg5b733oiMP",
        "ethNodeUrl": "https://data-seed-prebsc-1-s1.binance.org:8545",
        "ethMasterSk": "0x2bdd21761a483f71054e14f5b827213567971c676928d9a1808cbfa4b7501200",
        "nearClientValidateHeader": "true", 
        "nearClientValidateHeaderMode": "bsc",
        "nearClientSk": "ed25519:3D4YudUQRE39Lc4JHghuB5WM8kbgDDa34mnrEP5DdTApVH81af7e2dWgNPEaiQfdJnZq1CNPp5im4Rg5b733oiMP",
        "nearProverSk": "ed25519:3D4YudUQRE39Lc4JHghuB5WM8kbgDDa34mnrEP5DdTApVH81af7e2dWgNPEaiQfdJnZq1CNPp5im4Rg5b733oiMP",
        "ethEd25519Address": "0x62ff99964d1026bfd3fea2086b9d141917f1a793",
        "ethClientLockEthAmount": "1000",
        "ethClientLockDuration": "10",
        "ethClientAddress": "0x2cdaaa435937ae65705b97fbd01d28f4f6494545",
        "ethProverAddress": "0x5cdc96ce31bdf08ef39d5f65a9fd5e236ce2245f",
        "ethErc20Address": "0x7ef88a3901494435700ccc4c7615903b43ec2dbc",
        "ethLockerAddress": "0x24d718f7d8af7d7a4abaa2b1cb7abbb36a23a00d",
        "ethAdminAddress": "0xDf08F82De32B8d460adbE8D72043E3a7e25A3B39"
    }
```

# Deploy the bridge locally with BSC Mode

To run the bridge locally with Binance Smart Chain, run the following cmd: 

1. init yarn and generate Ethereum contracts: 

```bash
make init
```

2. Start Near blockchain and connect with Binance testnet.
```bash
make start-bsc
```

3. Deploy Near and Ethereum contracts

```bash
make full-bsc-contracts
```
or use already deployed contracts (on BSC), first copy past the following parameters inside the `~/.rainbow/config.json` file: 

```bash
    "ethEd25519Address": "0x62ff99964d1026bfd3fea2086b9d141917f1a793",
    "ethClientLockEthAmount": "1000",
    "ethClientLockDuration": "10",
    "ethClientAddress": "0x2cdaaa435937ae65705b97fbd01d28f4f6494545",
    "ethProverAddress": "0x5cdc96ce31bdf08ef39d5f65a9fd5e236ce2245f",
    "ethErc20Address": "0x7ef88a3901494435700ccc4c7615903b43ec2dbc",
    "ethLockerAddress": "0x24d718f7d8af7d7a4abaa2b1cb7abbb36a23a00d",
    "ethAdminAddress": "0xDf08F82De32B8d460adbE8D72043E3a7e25A3B39"
```
then run the following command:

```bash
make light-bsc-contracts
```

4. Start the relayers:

```bash
make start-relayer
```

5. When you are done you can stop the bridge by running:

```bash
make stop-all
```

- The bridge uses pm2 tool so you can use it to check logs and list the processes
- If you need BNB you can request some using [Binance faucet](https://testnet.binance.org/faucet-smart) using the following Address: `0xDf08F82De32B8d460adbE8D72043E3a7e25A3B39`

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+stop-all:
+	cli/index.js stop all
+
+build-eth-client:
+	cd contracts/near/eth-client && sudo ./build.sh
+
+test-eth-client:
+	cd contracts/near/eth-client && ./test.sh
+
+.PHONEY: help init start-ethash start-bsc gen-contracts deploy-contracts start-relayer stop-all build-eth-client
```
Comment (idirall22) [inline Makefile:61]: fixed

File: cli/commands/start/binance-smart-chain.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+// Binance Smart chain configurations to connect with testnet
```
Comment (idirall22) [inline cli/commands/start/binance-smart-chain.js:1]: fixed I added the possibility to use cmd args

File: contracts/near/eth-client/src/lib.rs [hunk 3]
```diff
@@ -348,12 +383,192 @@ impl EthClient {
...
 
-    /// Verify PoW of the header.
     fn verify_header(
         &self,
         header: &BlockHeader,
         prev: &BlockHeader,
         dag_nodes: &[DoubleNodeWithMerkleProof],
+    ) -> bool {
+        // if the validate_header is false, header are not validated. 
+        if self.validate_header == false {
```
[the remark this answers is not in the document]
Comment (3alpha) [reply contracts/near/eth-client/src/lib.rs]: You can take a look [here](https://ethereum.github.io/yellowpaper/paper.pdf) equation (48). That is the way how miners can expand and shrink block depending on the need. In POSA they just use fixed 60M gas limit.

File: contracts/near/eth-client/build.sh [hunk 4]
```diff
@@ -12,6 +12,7 @@ else
      userflag=""
 fi
 
+# simple:latest \
```
Comment (3alpha) [inline contracts/near/eth-client/build.sh:15]: You can remove this change to keep diff smaller

Comment (idirall22) [reply contracts/near/eth-client/build.sh:15]: ok

File: contracts/near/eth-client/src/lib.rs [hunk 5]
```diff
@@ -406,81 +410,65 @@ impl EthClient {
...
         let prev_gas_limit = format!("{}", prev.gas_limit).parse::<i64>().unwrap();
         let header_gas_limit = format!("{}", header.gas_limit).parse::<i64>().unwrap();
-
-        // Verify that the gas limit remains within allowed bounds
         let diff = (prev_gas_limit-header_gas_limit).abs();
         let limit = prev_gas_limit / gas_limit_bound_divisor;
-
+        
+        // Verify that the gas limit remains within allowed bounds
         if diff >= limit || header.gas_limit < U256(min_gas_limit.into()) {
```
Comment (3alpha) [inline contracts/near/eth-client/src/lib.rs:459]: Check for minimum gas is already done in basic verification, this is duplicate.

File: contracts/near/eth-client/src/lib.rs [hunk 6]
```diff
@@ -647,20 +623,21 @@ impl EthClient {
         (H256(pair.0), H256(pair.1))
     }  
 
-    // cache the bsc validators.
-    fn get_bsc_validators(header: &BlockHeader) -> UnorderedMap<Address, bool>{
+    // check if the author is in the validators set.
+    fn is_validator_exists (&self, epoch_header: &BlockHeader, add: Address) -> bool{
```
Comment (3alpha) [inline contracts/near/eth-client/src/lib.rs:627]: Rename to `validator_exists` or `is_valid_validator` or `is_in_validator_set`.

File: Makefile [hunk 7]
```diff
@@ -18,15 +18,20 @@ gen-contracts:
 	cd contracts/eth/nearbridge/ && yarn && yarn build
 	cd contracts/eth/nearprover/ && yarn && yarn build
 	
+# start near blockchain and connect with ganache.
+start-ethash:
+	cli/index.js clean
+	cli/index.js prepare --core-src ${HOME}/Desktop/core
```
Comment (3alpha) [inline Makefile:24]: Maybe we should change this absolute path to relative? Or in some special folder directly in the homedir?

**Pull request merged**